src/Compilers/Core/Portable/CodeAnalysisEventSource.cs

```diff
@@ -38,12 +37,95 @@ public static class Tasks
         internal void StartSingleGeneratorRunTime(string generatorName, string assemblyPath, string id) => WriteEvent(3, generatorName, assemblyPath, id);
 
         [Event(4, Message = "Generator {0} ran for {2} ticks", Keywords = Keywords.Performance, Level = EventLevel.Informational, Opcode = EventOpcode.Stop, Task = Tasks.SingleGeneratorRunTime)]
-        internal void StopSingleGeneratorRunTime(string generatorName, string assemblyPath, long elapsedTicks, string id) => WriteEvent(4, generatorName, assemblyPath, elapsedTicks, id);
+        internal void StopSingleGeneratorRunTime(string generatorName, string assemblyPath, long elapsedTicks, string id)
+        {
+            unsafe
+            {
+                fixed (char* generatorNameBytes = generatorName)
+                fixed (char* assemblyPathBytes = assemblyPath)
+                fixed (char* idBytes = id)
+                {
+                    Span<EventData> data = stackalloc EventData[]
+                    {
+                        GetEventDataForString(generatorName, generatorNameBytes),
+                        GetEventDataForString(assemblyPath, assemblyPathBytes),
+                        GetEventDataForLong(&elapsedTicks),
+                        GetEventDataForString(id, idBytes),
+                    };
+
+                    fixed (EventSource.EventData* dataPtr = data)
+                    {
+                        WriteEventCore(eventId: 4, data.Length, dataPtr);
+                    }
+                }
+            }
+        }
 
         [Event(5, Message = "Generator '{0}' failed with exception: {1}", Level = EventLevel.Error)]
         internal void GeneratorException(string generatorName, string exception) => WriteEvent(5, generatorName, exception);
 
         [Event(6, Message = "Node {0} transformed", Keywords = Keywords.Correctness, Level = EventLevel.Verbose, Task = Tasks.BuildStateTable)]
-        internal void NodeTransform(int nodeHashCode, string name, string tableType, int previousTable, string previousTableContent, int newTable, string newTableContent, int input1, int input2) => WriteEvent(6, nodeHashCode, name, tableType, previousTable, previousTableContent, newTable, newTableContent, input1, input2);
+        internal void NodeTransform(int nodeHashCode, string name, string tableType, int previousTable, string previousTableContent, int newTable, string newTableContent, int input1, int input2)
+        {
+            unsafe
```

Comment:
💡 Two more suggestions:

1. Move `unsafe` modifier to method
2. Change this line to `if (IsEnabled())`